Enable expiration of data in DynamoDB

mindstorms6 opened the pull request:
Howdy!
This is a work in progress for enabling expiration of historic data in the DDB Persistence side of OpenHab.  DDB has a built in [feature that expires items](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/TTL.html) when a ttl is present. My thought process was to

1. Ensure we create tables with a TTL (if enabled)
1. For data that is inserted into DDB - add an `expiration` (epoch seconds) field to it with a 90 day (but configurable) expiration date from the time of data collection.

It's not ready to merge just yet - but I wanted to get feedback from folks in the know (first pull request 😬)

Let me know the obvious things I missed / things you want to see.

Thanks!

mindstorms6 committed:
WIP: Enable expiration of data in DynamoDB

Signed-off-by: 🌈 💻  Miley <breland@bdawg.org>

mindstorms6 commented:
cc @ssalonen

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/AbstractDynamoDBItem.java:
```diff
@@ -212,4 +215,14 @@ public String toString() {
 
+    @Override
+    @DynamoDBAttribute(attributeName = ATTRIBUTE_NAME_EXPIRATION)
+    public Long getExpiration() {
```
I think you need to have getters and setters in subclasses (`DynamoDBBigDecimalItem` & `DynamoDBStringItem`), not in the parent class in order AWS SDK reflection to pick it up correctly. At least previously I encountered clear issues otherwise.

In any case, you can run the integration tests against real DynamoDB, see the instructions in [`.test` bundle `README`](https://github.com/openhab/openhab1-addons/blob/master/bundles/persistence/org.openhab.persistence.dynamodb.test/README.md)

ssalonen replied:
How about renaming it to more verbose `expirationDateEpochSeconds` to remove ambiguity with epoch-seconds and epoch-milliseconds?

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBItem.java:
```diff
@@ -48,6 +50,10 @@
 
+    Long getExpiration();
+
+    void setExpiration(long expirationTime);
```
Please rename getter/setter and setter argument name to make it clear whether we have milliepoch or epoch-seconds

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBPersistenceService.java:
```diff
@@ -363,6 +371,12 @@ protected void flushBufferedData() {
                 Deque<DynamoDBItem<?>> batch = entry.getValue();
+                if (dbConfig.isAutoExpirationEnabled()){
+                    for (DynamoDBItem<?> batchItem: batch){
+                        Instant expiration = batchItem.getTime().toInstant().plus(dbConfig.getAutoExpirationDays(), ChronoUnit.DAYS);
```
`toInstant` not allowed I believe... This repo is still pre-java8?

mindstorms6 replied:
I'll look for a response from our maintainer friends - I poked around the various `travis.yml` scripts (odd since we use jenkins?) And all the `pom.xml` files I could find and didn't see anything obvious that limits us to features lower than 8. Could totally have skipped something though, still a new code base to me.

In any event - we have `joda` in the `lib` so I assume I can just use joda time for this? 
If you'd like me to do so, just let me know.

ssalonen replied:
As discussed in comments, java 8 is ok -- no need to change anything :+1:

ssalonen commented on features/openhab-addons-external/src/main/resources/conf/dynamodb.cfg:
```diff
@@ -47,5 +47,12 @@
 # write capacity for the created tables
 #writeCapacityUnits=1
 
+# if auto expiration from ddb is enabled (ttl) [ defaults to false ]
```
Use DynamoDB instead previously unmentioned abbreviation `ddb`. Since other parameters do not have `[ defaults to XX ]` neither should this one. The default is there in the comments.

ssalonen commented on features/openhab-addons-external/src/main/resources/conf/dynamodb.cfg:
```diff
@@ -47,5 +47,12 @@
+# if auto expiration from ddb is enabled (ttl) [ defaults to false ]
+#autoExpirationEnabled=false
+
+# if auto expiration is enabled, nubmer of days to keep data for before expiring
```
typo nubmer

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBPersistenceService.java:
```diff
@@ -275,7 +271,19 @@ private synchronized boolean createTable(DynamoDBMapper mapper, Class<?> dtoClas
+                                )
+                );
+            }
+            return tableActive;
```
A question with implementation details -- is it possible that table is not active since auto-expiration parameter was just modified?

ssalonen replied:
Please see this comment @mindstorms6

ssalonen replied:
Please see this comment @mindstorms6

ssalonen reviewed (changes requested):
Looks good overall, couple of additional comments in addition to line comments

- please document new parameters in [README.md](https://github.com/openhab/openhab1-addons/blob/2c1d2a58dd2e7886595eaaad062b505fcb1ccd4a/bundles/persistence/org.openhab.persistence.dynamodb/README.md)
- please run integration tests using real DynamoDB -- instructions in `.test` bundle.

ssalonen commented:
@mindstorms6 Great to see new contributor to the DynamoDB addon, welcome! Please see my comments above

ssalonen commented:
Also, there are some repo changes coming towards persistence addons: 
https://github.com/openhab/openhab1-addons/pull/5844
https://github.com/openhab/openhab2-addons/pull/527.
https://github.com/openhab/openhab-core/pull/680

Once this is done, you might need to rebase the changes to the repo worked on in those PRs.

Actually, this also means that java 8 is probably ok at least in the new repo structure.

kaikreuzer commented:
> Actually, this also means that java 8 is probably ok at least in the new repo structure.

Yes, I agree - as the persistence add-ons are about to be moved after the 2.5 release, it should be fine (and actually much preferred) to use Java8 instead of Joda.

mindstorms6 committed:
CR feedback

Signed-off-by: 🌈 💻  Miley <breland@bdawg.org>

mindstorms6 committed:
Add integration tests, move get and set for expiration to child classes

Signed-off-by: 🌈 💻  Miley <breland@bdawg.org>

mindstorms6 commented:
I've addressed the above comment I believe, and added/run the integration tests. You were right - the `get/set` must be in each child class - TIL.

Marking as "ready for review"

Thanks!

ssalonen commented:
I get the below error with `ExpiringStringItemIntegrationTest` when running integration tests. Existing integration tests seem to pass.

Could you look into this?

```
com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMappingException: DynamoDBStringItem[expiration]; could not unconvert attribute
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperTableModel.unconvert(DynamoDBMapperTableModel.java:271)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper.privateMarshallIntoObject(DynamoDBMapper.java:456)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper.marshallIntoObjects(DynamoDBMapper.java:484)
	at com.amazonaws.services.dynamodbv2.datamodeling.PaginatedQueryList.<init>(PaginatedQueryList.java:65)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper.query(DynamoDBMapper.java:1506)
	at com.amazonaws.services.dynamodbv2.datamodeling.AbstractDynamoDBMapper.query(AbstractDynamoDBMapper.java:265)
	at org.openhab.persistence.dynamodb.internal.DynamoDBPersistenceService.query(DynamoDBPersistenceService.java:501)
	at org.openhab.persistence.dynamodb.internal.AbstractTwoItemIntegrationTest.testQueryUsingNameAndEnd(AbstractTwoItemIntegrationTest.java:160)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
Caused by: com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMappingException: could not invoke null on class org.openhab.persistence.dynamodb.internal.DynamoDBStringItem with value 1565326414 of type class java.lang.Long
	at com.amazonaws.services.dynamodbv2.datamodeling.StandardBeanProperties$MethodReflect.set(StandardBeanProperties.java:136)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperFieldModel.set(DynamoDBMapperFieldModel.java:111)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperFieldModel.unconvertAndSet(DynamoDBMapperFieldModel.java:164)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperTableModel.unconvert(DynamoDBMapperTableModel.java:267)
	... 31 more
Caused by: java.lang.NullPointerException
	at com.amazonaws.services.dynamodbv2.datamodeling.StandardBeanProperties$MethodReflect.set(StandardBeanProperties.java:133)
	... 34 more

```

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/README.md:
```diff
@@ -98,7 +98,8 @@ In addition to the configuration properties above, the following are also availa
-
+| autoExpirationEnabled      | false      |    No    | If historical events should be removed using DyanmoDB's TTL feature. This avoids infinite growth.  |
+| autoExpirationDays         | 90         |    No    | If autoExpirationEnabled is true - how long should historical events be kept for, in days.         |
 Typically you should not need to modify parameters related to buffering. 
```
Add line break to fix formatting.

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/ExpiringStringItemIntegrationTest.java:
```diff
@@ -0,0 +1,60 @@
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class ExpiringStringItemIntegrationTest extends StringItemIntegrationTest {
```
Test fails:

e.g. `testQueryUsingName`
```
com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMappingException: DynamoDBStringItem[expiration]; could not unconvert attribute
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperTableModel.unconvert(DynamoDBMapperTableModel.java:271)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper.privateMarshallIntoObject(DynamoDBMapper.java:456)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper.marshallIntoObjects(DynamoDBMapper.java:484)
	at com.amazonaws.services.dynamodbv2.datamodeling.PaginatedQueryList.<init>(PaginatedQueryList.java:65)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper.query(DynamoDBMapper.java:1506)
	at com.amazonaws.services.dynamodbv2.datamodeling.AbstractDynamoDBMapper.query(AbstractDynamoDBMapper.java:265)
	at org.openhab.persistence.dynamodb.internal.DynamoDBPersistenceService.query(DynamoDBPersistenceService.java:501)
	at org.openhab.persistence.dynamodb.internal.AbstractTwoItemIntegrationTest.testQueryUsingName(AbstractTwoItemIntegrationTest.java:131)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
Caused by: com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMappingException: could not invoke null on class org.openhab.persistence.dynamodb.internal.DynamoDBStringItem with value 1565326414 of type class java.lang.Long
	at com.amazonaws.services.dynamodbv2.datamodeling.StandardBeanProperties$MethodReflect.set(StandardBeanProperties.java:136)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperFieldModel.set(DynamoDBMapperFieldModel.java:111)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperFieldModel.unconvertAndSet(DynamoDBMapperFieldModel.java:164)
	at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperTableModel.unconvert(DynamoDBMapperTableModel.java:267)
	... 31 more
Caused by: java.lang.NullPointerException
	at com.amazonaws.services.dynamodbv2.datamodeling.StandardBeanProperties$MethodReflect.set(StandardBeanProperties.java:133)
	... 34 more
```

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBBigDecimalItem.java:
```diff
@@ -81,6 +87,9 @@ public void setTime(Date time) {
     }
 
+    @Override
+    public void setExpirationDateEpochSeconds(long expirationDateEpochSeconds) { this.expirationDateEpochSeconds = expirationDateEpochSeconds; }
```
Format source code using eclipse

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBStringItem.java:
```diff
@@ -72,4 +78,7 @@ public void setState(String state) {
     }
+
+    @Override
+    public void setExpirationDateEpochSeconds(long expirationDateEpochSeconds) { this.expirationDateEpochSeconds = expirationDateEpochSeconds; }
```
Format code using eclipse

ssalonen reviewed (changes requested):
Please see the comments

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/README.md:
```diff
@@ -113,6 +114,8 @@ When an item is persisted via this service, a table is created (if necessary). C
 
 Each table will have three columns: `itemname` (item name), `timeutc` (in ISO 8601 format with millisecond accuracy), and `itemstate` (either a number or string representing item state).
 
+If `autoExpiration` is enabled, the tables will have the TTL Deletion feature turned on and an addtitional column called `expiration`. (numeric type) 
```
typo addtitional

kaikreuzer commented:
@mindstorms6 Are you still working on this PR?

mindstorms6 commented:
Slipped my mind - but I’ll get back on it this weekend with the requested
changes. My apologies for the delay.

On Sun, Dec 8, 2019 at 1:25 PM Kai Kreuzer <notifications@github.com> wrote:

> @mindstorms6 <https://github.com/mindstorms6> Are you still working on
> this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab1-addons/pull/5875?email_source=notifications&email_token=AAAWWCKLONZSCIEMOZECSZTQXVQ37A5CNFSM4H4YYNYKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGHKGSA#issuecomment-562996040>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAWWCIQDIEM4HLQ4PWJHRLQXVQ37ANCNFSM4H4YYNYA>
> .
>
